phoenix-core/src/main/java/org/apache/phoenix/monitoring/connqueryservice/ConnectionQueryServicesMetricsManager.java

```diff
@@ -0,0 +1,375 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+/**
+ * Central place where we keep track of all the Connection Query Service metrics. Register each
+ * Connection Query Service and store the instance of it associated with ConnectionServiceName in a
+ * map This class exposes following functions as static methods to help catch all exception
+ * 1.clearConnectionQueryServiceMetrics
+ * 2.getTableMetricsMethod
+ * 3.pushMetricsFromConnInstanceMethod
+ * 4.updateMetricsMethod
+ */
+public class ConnectionQueryServicesMetricsManager {
+    private static final Logger LOGGER =
+            LoggerFactory.getLogger(ConnectionQueryServicesMetricsManager.class);
+    private static boolean isConnectionQueryServiceMetricsEnabled;
+    private static boolean isConnectionQueryServiceMetricPublisherEnabled;
+    private static ConcurrentMap<String, PhoenixConnectionQueryServiceMetrics>
+            connectionQueryServiceMetricsMapping;
+    // Singleton object
+    private static volatile ConnectionQueryServicesMetricsManager
+            connectionQueryServicesMetricsManager = null;
+    private static volatile MetricPublisherSupplierFactory mPublisher = null;
+    private static volatile QueryServicesOptions options;
+
+    @VisibleForTesting
+    public ConnectionQueryServicesMetricsManager(QueryServicesOptions opts) {
+        options = opts;
+        isConnectionQueryServiceMetricsEnabled = options.isConnectionQueryServiceMetricsEnabled();
+        connectionQueryServiceMetricsMapping = new ConcurrentHashMap<>();
+        LOGGER.info("Phoenix connection query service metrics enabled status: "
+                + isConnectionQueryServiceMetricsEnabled);
+        // Return simply if connection query service metrics are not enabled
+        if (!isConnectionQueryServiceMetricsEnabled) {
+            return;
+        }
+        isConnectionQueryServiceMetricPublisherEnabled =
+                options.isConnectionQueryServiceMetricsPublisherEnabled();
+        LOGGER.info("Phoenix connection query service metrics publisher enabled status: "
+                + isConnectionQueryServiceMetricPublisherEnabled);
+    }
+
+    @VisibleForTesting
+    public static void setInstance(ConnectionQueryServicesMetricsManager metricsManager) {
+        connectionQueryServicesMetricsManager = metricsManager;
+    }
+
+    /**
+     * Method to provide instance of ConnectionQueryServiceMetricsManager(Create if needed in
+     * thread safe manner)
+     * @return returns instance of ConnectionQueryServicesMetricsManager for the said
+     *         ConnectionQueryService
+     */
+    public static ConnectionQueryServicesMetricsManager getInstance() {
+        ConnectionQueryServicesMetricsManager localRef = connectionQueryServicesMetricsManager;
```

Comment:
localRef doesn't seem to be required right? this code can work on connectionQueryServicesMetricsManager ?

Reply:
Done. Removed use of `localRef`